runtime/grid_sync.cu

```diff
@@ -150,4 +150,82 @@ __device__ void sync(
   block_sync::sync<Aligned>();
 }
 
+// Non-blocking function to acquire the semaphore value in each calling thread
+__device__ int64_t semaphoreFetch(int64_t* semaphore) {
+  int64_t state;
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 700
+  asm volatile("ld.global.acquire.gpu.b64 %0, [%1];\n"
+               : "=l"(state)
+               : "l"(semaphore));
+#else
+  asm volatile("ld.global.cg.b64 %0, [%1];\n" : "=l"(state) : "l"(semaphore));
+#endif
+  return state;
+}
+
+// Sync block then set semaphore to new_value
+__device__ void semaphoreRelease(int64_t* semaphore, int64_t new_value) {
+  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0) {
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 700
+    asm volatile("st.global.release.gpu.b64 [%0], %1;\n"
+                 :
+                 : "l"(semaphore), "l"(new_value));
+#else
+    asm volatile("st.global.cg.b64 [%0], %1;\n"
```

Comment:
Maybe `.cg` here is important as we want to make sure this write is visible to the other blocks.